File: csharp/config/store/Anchor.cs
/* 
 Copyright (c) 2010, Direct Project
 All rights reserved.

 Authors:
    Umesh Madan     umeshma@microsoft.com
  
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of The Direct Project (directproject.org) nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
*/
using System;
using System.Data.Linq.Mapping;
using System.Security.Cryptography.X509Certificates;
using System.Runtime.Serialization;

using Health.Direct.Common.Extensions;

namespace Health.Direct.Config.Store
{
    [Table(Name = "Anchors")]
    [DataContract(Namespace = ConfigStore.Namespace)]
    public class Anchor
    {
        public const int MaxOwnerLength = 400;
        
        string m_owner;
        byte[] m_data;
        
        public Anchor()
        {
        }

        public Anchor(string owner, byte[] sourceFileBytes, string password)
            : this(owner, Certificate.Import(sourceFileBytes, password))
        {
        }
        
        
        public Anchor(string owner, X509Certificate2 certificate)
            : this(owner, certificate, true, true)
        {
        }
        
        public Anchor(string owner, X509Certificate2 certificate, bool forIncoming, bool forOutgoing)
        {
            if (certificate == null)
            {
                throw new ConfigStoreException(ConfigStoreError.InvalidCertificate);
            }
            
            this.Owner = owner;
            this.Thumbprint = certificate.Thumbprint;
            this.Data = certificate.RawData;
            this.CreateDate = DateTimeHelper.Now;
            this.ValidStartDate = certificate.NotBefore;
            this.ValidEndDate = certificate.NotAfter;
            this.ForIncoming = forIncoming;
            this.ForOutgoing = forOutgoing;
        }

        [Column(Name = "Owner", CanBeNull = false, IsPrimaryKey = true)]
        [DataMember(IsRequired = true)]
        public string Owner
        {
            get
            {
                return m_owner;
            }
            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ConfigStoreException(ConfigStoreError.InvalidOwnerName);
                }
                
                if (value.Length > MaxOwnerLength)
                {
                    throw new ConfigStoreException(ConfigStoreError.OwnerLength);
                }
                
                m_owner = value;
            }
        }

        [Column(Name = "Thumbprint", CanBeNull = false, IsPrimaryKey = true)]
        [DataMember(IsRequired = true)]
        public string Thumbprint
        {
            get;
            set;
        }

        [Column(Name = "CertificateID", CanBeNull = false, IsDbGenerated=true, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public long ID
        {
            get;
            set;
        }
        
        [Column(Name = "CertificateData", DbType = "varbinary(MAX)", CanBeNull = false, UpdateCheck = UpdateCheck.Never)]
        [DataMember(IsRequired = true)]
        public byte[] Data
        {
            get
            {
                return m_data;
            }
            set
            {
                m_data = value;
            }
        }

        [Column(Name = "CreateDate", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public DateTime CreateDate
        {
            get;
            set;
        }

        [Column(Name = "ValidStartDate", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public DateTime ValidStartDate
        {
            get;
            set;
        }

        [Column(Name = "ValidEndDate", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public DateTime ValidEndDate
        {
            get;
            set;
        }

        [Column(Name = "ForIncoming", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public bool ForIncoming
        {
            get;
            set;
        }

        [Column(Name = "ForOutgoing", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public bool ForOutgoing
        {
            get;
            set;
        }

        [Column(Name = "Status", CanBeNull = false, UpdateCheck = UpdateCheck.WhenChanged)]
        [DataMember(IsRequired = true)]
        public EntityStatus Status
        {
            get;
            set;
        }
        
        public bool HasData
        {
            get
            {
                return (this.Data != null && this.Data.Length > 0);
            }
        }                              

        public bool IsValid(DateTime testDate)
        {
            return (this.ValidStartDate <= testDate && testDate <= this.ValidEndDate);
        }

        public void ClearData()
        {
            m_data = null;
        }

        public void ValidateHasData()
        {
            if (m_data.IsNullOrEmpty())
            {
                throw new ConfigStoreException(ConfigStoreError.MissingCertificateData);
            }
        }
        
        public X509Certificate2 ToX509Certificate()
        {
            return new X509Certificate2(this.Data);
        }
        
        public static X509Certificate2Collection ToX509Collection(Anchor[] anchors)
        {
            if (anchors.IsNullOrEmpty())
            {
                return null;
            }
            
            X509Certificate2Collection x509Collection = new X509Certificate2Collection();
            if (anchors != null)
            {
                for (int i = 0; i < anchors.Length; ++i)
                {
                    x509Collection.Add(anchors[i].ToX509Certificate());
                }
            }            
            return x509Collection;
        }
    }
}

File: java/config/config-model/src/main/java/org/nhindirect/config/model/Anchor.java
/* 
Copyright (c) 2010, NHIN Direct Project
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer 
   in the documentation and/or other materials provided with the distribution.  
3. Neither the name of the The NHIN Direct Project (nhindirect.org) nor the names of its contributors may be used to endorse or promote 
   products derived from this software without specific prior written permission.
   
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS 
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE 
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.nhindirect.config.model;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.Calendar;

import org.codehaus.enunciate.json.JsonRootType;
import org.codehaus.jackson.annotate.JsonIgnore;
import org.nhindirect.config.model.exceptions.CertificateConversionException;


/**
 * A trust anchor used for discerning trust of an end entity certificates.
 * @author Greg Meyer
 * @since 1.0
 */
///CLOVER:OFF
@JsonRootType
public class Anchor 
{
    private String owner;
    private String thumbprint;
    private long certificateId;
    private byte[] certificateData;
    private long id;
    private Calendar createTime;
    private Calendar validStartDate;
    private Calendar validEndDate;
    private EntityStatus status;
    private boolean incoming;
    private boolean outgoing;
    
    /**
     * Empty constructor.
     */
    public Anchor()
    {
    	
    }

    /**
     * Gets the owner of the anchor.  The owner is domain name that this anchor is associate with.
     * @return The owner of the anchor.
     */
	public String getOwner() 
	{
		return owner;
	}

	/**
	 * Sets the owner of the anchor.
	 * @param owner The owner of the anchor.
	 */
	public void setOwner(String owner) 
	{
		this.owner = owner;
	}

	/**
	 * Gets the thumbprint of the anchor.  A thumbprint is a SHA-1 has of the DER encoding of the anchor.
	 * @return The thumbprint of the anchor.
	 */
	public String getThumbprint() 
	{
		return thumbprint;
	}

	/**
	 * Sets the thumbprint of the anchor.
	 * @param thumbprint The thumbprint of the anchor.
	 */
	public void setThumbprint(String thumbprint) 
	{
		this.thumbprint = thumbprint;
	}

	/**
	 * Gets the internal id of the anchor.
	 * @return The interal id of the anchor.
	 */
	public long getCertificateId() 
	{
		return certificateId;
	}

	/**
	 * Sets the internal id of the anchor.
	 * @param certificateId The internal id of the anchor.
	 */
	public void setCertificateId(long certificateId) 
	{
		this.certificateId = certificateId;
	}

	/**
	 * Gets the DER encoded data of the anchor.
	 * @return The DER encoded data of the anchor.
	 */
	public byte[] getCertificateData() 
	{
		return certificateData;
	}

	/**
	 * Sets the DER encoded data of the anchor.
	 * @param certificateData The DER encoded data of the anchor.
	 */
	public void setCertificateData(byte[] certificateData) 
	{
		this.certificateData = certificateData;
	}

	/**
	 * Gets the internal id of the anchor.
	 * @return The interal id of the anchor.
	 */
	public long getId() 
	{
		return id;
	}

	/**
	 * Sets the internal id of the anchor.
	 * @param certificateId The internal id of the anchor.
	 */
	public void setId(long id) 
	{
		this.id = id;
	}

	/**
	 * Gets the date/time that the anchor was added to the system.
	 * @return The date/time that the anchor was added to the system.
	 */
	public Calendar getCreateTime() 
	{
		return createTime;
	}

	/**
	 * Sets the date/time that the anchor was added to the system.
	 * @param createTime The date/time that the anchor was added to the system.
	 */
	public void setCreateTime(Calendar createTime) 
	{
		this.createTime = createTime;
	}

	/**
	 * Gets the valid from date/time of the anchor.
	 * @return The valid from date/time of the anchor.
	 */
	public Calendar getValidStartDate() 
	{
		return validStartDate;
	}

	/** 
	 * Sets the valid from date/time of the anchor.
	 * @param validStartDate The valid from date/time of the anchor.
	 */
	public void setValidStartDate(Calendar validStartDate) 
	{
		this.validStartDate = validStartDate;
	}

	/**
	 * Gets the valid until date/time of the anchor.
	 * @return The valid until date/time of the anchor.
	 */
	public Calendar getValidEndDate() 
	{
		return validEndDate;
	}

	/**
	 * Sets the valid until date/time of the anchor.
	 * @param validEndDate The valid until date/time of the anchor.
	 */
	public void setValidEndDate(Calendar validEndDate) 
	{
		this.validEndDate = validEndDate;
	}

	/**
	 * Gets the status of the anchor.
	 * @return The status of the anchor.
	 */
	public EntityStatus getStatus() 
	{
		return status;
	}

	/**
	 * Sets the status of the anchor.
	 * @param status The status of the anchor.
	 */
	public void setStatus(EntityStatus status) 
	{
		this.status = status;
	}

	/**
	 * Determines if the anchor is valid for incoming messages.
	 * @return True if the anchor can be use for incoming messages; false otherwise.
	 */
	public boolean isIncoming() 
	{
		return incoming;
	}

	/**
	 * Sets if the anchor is valid for incoming messages.
	 * @param incoming True if the anchor can be use for incoming messages; false otherwise.
	 */
	public void setIncoming(boolean incoming) 
	{
		this.incoming = incoming;
	}

	/**
	 * Determines if the anchor is valid for outgoing messages.
	 * @return True if the anchor can be use for outgoing messages; false otherwise.
	 */
	public boolean isOutgoing() 
	{
		return outgoing;
	}

	/**
	 * Sets if the outgoing is valid for outgoing messages.
	 * @param outgoing True if the anchor can be use for outgoing messages; false otherwise.
	 */
	public void setOutgoing(boolean outgoing) 
	{
		this.outgoing = outgoing;
	}
    
	@JsonIgnore
	/**
	 * The returned value is derived from the internal byte stream representation.  This attribute is suppressed during JSON conversion.
	 */
	public X509Certificate getAnchorAsX509Certificate()
	{
        final InputStream bais = new ByteArrayInputStream(certificateData);

        try
        {
        	return (X509Certificate) CertificateFactory.getInstance("X.509").generateCertificate(bais);
        } 
        catch (CertificateException e) 
        {
        	throw new CertificateConversionException(e);
		}
        finally
        {
        	try {bais.close();} catch (IOException ex) {}
        }
	}	
}
///CLOVER:ON
